s3/client.go

```diff
@@ -0,0 +1,424 @@
+	keys, err := s.iamClient.ListAccessKeys(&iam.ListAccessKeysInput{
+		UserName: aws.String(username),
+	})
+	if err != nil {
+		return err
+	}
+	if keys != nil {
+		for _, k := range keys.AccessKeyMetadata {
+			_, err := s.iamClient.DeleteAccessKey(&iam.DeleteAccessKeyInput{
+				UserName:    aws.String(username),
+				AccessKeyId: k.AccessKeyId,
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+	_, err = s.iamClient.DeleteUser(&iam.DeleteUserInput{
+		UserName: aws.String(username),
+	})
+	return err
+}
+
+func (s *S3Client) tagBucket(instanceID string, tags []*s3.Tag) (output *s3.PutBucketTaggingOutput, err error) {
+	createTagsInput := s3.PutBucketTaggingInput{
+		Bucket:  aws.String(s.buildBucketName(instanceID)),
+		Tagging: &s3.Tagging{TagSet: tags},
+	}
+	result, err := s.s3Client.PutBucketTagging(&createTagsInput)
+	return result, err
+}
+
+func (s *S3Client) buildBucketArns(fullBucketName string) (wildcardArn, bareArn string) {
+	return fmt.Sprintf("arn:aws:s3:::%s/*", fullBucketName), fmt.Sprintf("arn:aws:s3:::%s", fullBucketName)
+}
+
+func (s *S3Client) buildBucketName(bucketName string) string {
+	return fmt.Sprintf("%s%s", s.bucketPrefix, bucketName)
+}
+
+func (s *S3Client) AddUserToBucketPolicy(userArn, fullBucketName, policyDocument string) (PolicyDocument, error) {
+	updatedPolicy := BucketPolicyTemplate
+	updatedPolicy.Statement[0].Resource, updatedPolicy.Statement[1].Resource = s.buildBucketArns(fullBucketName)
+
+	if policyDocument == "" {
+		for policyStatementIndex := range updatedPolicy.Statement {
+			updatedPolicy.Statement[policyStatementIndex].Principal = Principal{
+				AWS: userArn,
+			}
+		}
+		return updatedPolicy, nil
+	}
+
+	err := json.Unmarshal([]byte(policyDocument), &updatedPolicy)
+	if err != nil {
+		return updatedPolicy, err
+	}
+	for policyStatementIndex := range updatedPolicy.Statement {
+		principal := updatedPolicy.Statement[policyStatementIndex].Principal.AWS
+		var existingPrincipals []interface{}
+		switch p := principal.(type) {
+		case string:
+			existingPrincipals = []interface{}{p}
+		case []interface{}:
+			existingPrincipals = p
+		}
+
+		var principals []string
+		for _, existingPrincipal := range existingPrincipals {
+			principals = append(principals, existingPrincipal.(string))
+		}
+		principals = append(
+			principals,
+			userArn,
+		)
+
+		updatedPolicy.Statement[policyStatementIndex].Principal.AWS = principals
+	}
+	return updatedPolicy, nil
+}
+
+func (s *S3Client) RemoveUserFromBucket(bindingID, bucketName string) error {
+	username := s.bucketPrefix + bindingID
+	fullBucketName := s.buildBucketName(bucketName)
+
+	getBucketPolicyOutput, err := s.s3Client.GetBucketPolicy(&s3.GetBucketPolicyInput{
+		Bucket: aws.String(fullBucketName),
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = s.RemoveUserFromBucketPolicy(username, fullBucketName, *getBucketPolicyOutput.Policy)
+	if err != nil {
+		return err
+	}
+
+	err = s.deleteUser(username)
```

Comment:
Perhaps fine in this context, but it's a bit surprising that a function called `RemoveUserFromBucket` actually ends up deleting the user - perhaps it would be better to rename it to make that clear? e.g. `RemoveUserFromBucketAndDeleteUser` or something.